How to enable reward sharing guide - Closes 1796 (#1901)
LiskHQ/lisk-docs: docs/modules/ROOT/pages/run-blockchain/reward-sharing.adoc

    ],
    "params": {
      "newCommission": 8500
    },
    "id": "7ad7d91488abdb859cf06570b17986a2ac14018584326d302afd0770fba96c86"
  }
}
Transaction with id: '7ad7d91488abdb859cf06570b17986a2ac14018584326d302afd0770fba96c86' received by node.
----
+
. Upon successful execution of the transaction, the validator's commission value will be updated.
To check the updated values, invoke the xref:{url_getvalidator}[pos_getValidator] endpoint.
+
[source,json]
----
{
    "name": "web3lord",
    "totalStake": "2011000000000",
    "selfStake": "1011000000000",
    "lastGeneratedHeight": 4260,
    "isBanned": false,
    "reportMisbehaviorHeights": [],
    "consecutiveMissedBlocks": 0,
    "commission": 8500,
    "lastCommissionIncreaseHeight": 271,
    "sharingCoefficients": [ //This object will automatically have a value, once the validator starts sharing rewards.
        {
            "tokenID": "1234567800000000",
            "coefficient": "0b67f3fb9ef253f78c1b2c"
        }
    ],
    "address": "lskm87us5hykopm2f2nxa92z5ftbr9r52kg5b45e6",
    "punishmentPeriods": []
}
----

== Claiming rewards

> After each block generated by the validator, both the validator and the staker(s) will earn a set percentage of rewards.

Reviewer: ```suggestion
After each block is generated, both the block generator and the staker(s) supporting them will earn a set percentage of rewards.

The part of rewards that is attributed to the validator that generated the block is automatically assigned to them by the Lisk Protocol. This is done by incrementing the `availableBalance` of the validator's account by the amount of rewards corresponding to the commission and self-stake.

On the other hand, stakers needs to manually claim their rewards, and that can be done via both Lisk Desktop and CLI. The part of rewards belonging to stakers gets locked in the validator's account (by increasing its `lockedBalance`) until each staker claims their rewards.
```

NOTE: Validators need to self-stake for their accounts and enable generation to become eligible to generate blocks.
As a result, validators receive additional rewards as per their self-stake, just like their other stakers do.
For example, consider an active validator has 0% commission, and at the same time, no other accounts are staking for the validator.
In this case, the validator would still receive 100% of the block rewards.

The reward-sharing mechanism is part of the PoS module. To learn more about how reward sharing works in Lisk Protocol, see {url_lip_70}[LIP 70^].

The commission values can range between *0* and *10000* (both inclusive), corresponding to percentages with two decimal places, which can be obtained by dividing the value by *100*. For example, a commission value of *1000*, corresponds to a *10%* commission.

Initially, after the validator's registration the commission of a validator is initialized with 100%.
Afterward, it can be decreased by submitting a commission change transaction.

> NOTE: A validator's commission can be decreased by any value, however, the commission can only be increased by 5% in a single change commission transaction.
> The validator has to then wait for the next commission increase period. For more information, see {url_constraints-on-commission-increase}[Constraints on Commission Increase^].

Author: ```suggestion
NOTE: A validator’s commission can be decreased at any time, by any value, however, the commission can only be increased by 5% in a single change commission transaction.
The validator has to wait for 28 days until they can increase the commission again.
For more information, see {url_constraints-on-commission-increase}[Constraints on Commission Increase^].
```

:url_chain_getGeneratorList: api/lisk-node-rpc.adoc#chain_getgeneratorlist
:url_getexpectedvalidatorrewards: api/module-rpc-api/dynamic-reward-endpoints.adoc#dynamicreward_getexpectedvalidatorrewards
:url_getvalidator: api/module-rpc-api/pos-endpoints.adoc#pos_getvalidator
:url_getclaimablerewards: api/module-rpc-api/pos-endpoints.adoc#pos_getclaimablerewards
:url_getbalance: api/module-rpc-api/token-endpoints.adoc#token_getbalance
:url_posting_transaction: integrate-blockchain/posting-transactions.adoc#posting-transaction-with-the-node-cli

:url_lip_70: https://github.com/LiskHQ/lips/blob/350d08a90bdbedb3485363a40dd34f3ccadf0b7d/proposals/lip-0070.md
:url_constraints-on-commission-increase: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0070.md#constraints-on-commission-increase
:url_reward_distribution: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0070.md#distribution-of-rewards

On this page, we'll talk about how a validator can enable reward sharing with their stakers and how the shared rewards can be later claimed by the stakers.

== What is reward sharing?

In the Lisk protocol, validators receive rewards for generating blocks.

> The set of validators that generate blocks, called generators, is chosen based on the total stake of each validator (self-stake and sum of staked amounts from stakers).
> 

Author: ```suggestion
```

= How to share block rewards
Muhammad Talha <muhammad.talha@lightcurve.io>
:toc: preamble
:toclevels: 5
:page-toclevels: 4
:idprefix:
:idseparator: -
:experimental:

:url_staking_guide: run-blockchain/staking.adoc
:url_getStaker: api/module-rpc-api/pos-endpoints.adoc#pos_getstaker
:url_consensus_getbftparameters: api/lisk-node-rpc.adoc#consensus_getbftparameters
:url_getexpectedvalidatorrewards: api/module-rpc-api/dynamic-reward-endpoints.adoc#dynamicreward_getexpectedvalidatorrewards
:url_getvalidator: api/module-rpc-api/pos-endpoints.adoc#pos_getvalidator
:url_getclaimablerewards: api/module-rpc-api/pos-endpoints.adoc#pos_getclaimablerewards
:url_getbalance: api/module-rpc-api/token-endpoints.adoc#token_getbalance
:url_posting_transaction: integrate-blockchain/posting-transactions.adoc#posting-transaction-with-the-node-cli

:url_lip_70: https://github.com/LiskHQ/lips/blob/350d08a90bdbedb3485363a40dd34f3ccadf0b7d/proposals/lip-0070.md

On this page, we'll talk about how a registered validator can enable reward sharing with its stakers and how such rewards can be later claimed.

== What is reward sharing?

In the Lisk protocol, validators receive rewards for generating blocks.

> The set of validators that generate blocks is chosen based on the total stake of each validator (self-stake and sum of staked amounts from stakers).

Reviewer 2: Don't know why this was resolved, the sentence is still there.

> The set of validators that generate blocks is chosen based on the total stake of each validator (self-stake and sum of staked amounts from stakers).

Author: The comment was resolved since another suggestion on the same line was accepted.

Rewards are shared between validators and stakers as follows: a certain part of the rewards is awarded to the validator as a commission.

> The precise percentage of the commission is chosen by each validator; then, the remaining rewards are shared by the validator and the stakers proportionally to their staked amounts.

Reviewer 2: ```suggestion
The precise percentage of the commission is chosen by each validator; then, the remaining rewards are shared among the stakers of the validator, proportionally to their staked amounts.

NOTE: Because validators need to self-stake for their accounts to become eligible to generate blocks, validators will receive additional rewards here, like their other stakers, too. For example, consider an active validator has 0% commission, and at the same time no other accounts are staking for the validator. In this case, the validator would still receive 100% of the block rewards.
```

The reward-sharing mechanism is part of the PoS module. To learn more about how reward sharing works in Lisk Protocol, see {url_lip_70}[LIP 70^].

The values of all constants related to commissions are between *0* and *10000*, corresponding to percentages with two decimal places, which can be obtained by dividing the value by *100*. For example, a commission value of *1000*, corresponds to a *10%* commission.

Initially, after the validator's registration the commission of a validator is initialized with 100%.

> Afterward, it can be decreased by submitting a commission change transaction. 

Reviewer 2: I would add a note below, explaining commission can be descreased by any value, but for increase it can only happen in steps

== Prerequisites and Background

* To earn rewards, a *staker* must stake LSKs for an active block-generating validator.
To learn more about staking, see xref:{url_staking_guide}[].

Once you successfully stake for a validator, you can see your stakes by invoking the xref:{url_getStaker}[pos_getStaker] endpoint.
Each time a validator becomes part of the block generation round, there are chances that the validator will be selected to generate an upcoming block. 

To check if the validator is in the list of active validators, look for the validator in the response of xref:{url_consensus_getbftparameters}[consensus_getBFTParameters].

> If the validator is in the returned list then, soon your validator will be generating a block and as a result will earn rewards.

Author: Alright, I have reported the bug: https://github.com/LiskHQ/lisk-docs/issues/1905

Block rewards are distributed after the `offset` block height is reached. By default, the value of offset is `2160`.

> After reaching the offset height, rewards for each validator are credited to their account.

Reviewer 2: This seems to be an edge case after a hardfork, I wouldn't mention it here

Reviewer 2: Sentence is still there, this should not have been resolved.

Author: IMHO this should stay, while I was performing the steps necessary for reward sharing, It was unclear to me for sometime as to why the rewards were not being shared. I later learned that there is this `offset` value and I think, for any new node or even with hard forks, this info adds value.

Reviewer 2: I see, but imho it is still out of scope for the guide, because this guide is targeted to validators(not sidechain developers), and they generally join an existing networks, so there is no offset to consider. The only time this is to consider is after a hardfork, which is a very special case, where consequences of the `offset` can be communicated seperatly.

Author: Fair point!

The estimated rewards of a validator can be checked via the xref:{url_getexpectedvalidatorrewards}[dynamicReward_getExpectedValidatorRewards] endpoint.

:sectnums:
== Sharing rewards

=== Sharing via Lisk Desktop
To share rewards via Lisk Desktop, the validator should be logged in to their account and then perform the following steps:

. Go to *My validator profile*, and click on the edit icon right next to the current `Commission` value:
+
image:run-blockchain/reward-sharing/share-reward-0.png["Click on edit commission",300]
+
. Type `85`, in the *Edit Commission* dialog box, and click on the *Confirm* button.

> In this example, the validator is set to keep *85%* of the rewards and share *15%* with their staker, hence the value of `85`.

Reviewer 2: Maybe we should use a low commission as an example, maybe even 0%, as we want to incentivize validators to set low commisions

Reviewer 2: Commission is still 85%, this should not have been resolved.

Author: I think the value 85 serves the purpose of this guide and there is not need to change it.

Reviewer 2: Afaik it was communicated to change commision of our testnet validators, to incentivize users to also choose lower commissions, too. Imho it would also help if the example in the guide doesn't choose a too high commission, for this purpose. 

But I can see 85 is used in the screenshots and it is cumbersome to change it now. I would suggest to add at least a TODO, to lower the commission in the future, if new screenshots need to be created anyway.

Author: TODO added!

    "fee": "1000000",
    "nonce": "10",
    "senderPublicKey": "65984ff3e6fe0d161a0a118c4e5d181a23c18f1d4bf59c78d178b7fcf0cadead",
    "signatures": [
      "f4e53ed1aaf56f878ab7cc13514164bbbe403dd52ac558ab88afdf8bc4829fd959aaf913ccf52e755185c681e13f84b4aaa723709faad7bc5de818fa96e48804"
    ],
    "params": {
      "newCommission": 8500
    },
    "id": "7ad7d91488abdb859cf06570b17986a2ac14018584326d302afd0770fba96c86"
  }
}
----
+
. Post the transaction to the node for its execution.
+
[tabs]
=====
Mainchain node::
+
--
[source,bash]
----
lisk-core transaction:send 0a03706f7312106368616e6765436f6d6d697373696f6e180a20c0843d2a2065984ff3e6fe0d161a0a118c4e5d181a23c18f1d4bf59c78d178b7fcf0cadead320308c03e3a40f4e53ed1aaf56f878ab7cc13514164bbbe403dd52ac558ab88afdf8bc4829fd959aaf913ccf52e755185c681e13f84b4aaa723709faad7bc5de818fa96e48804
----
--
Sidechain Node::
+
--
[source,bash]
----
./bin/run transaction:send 0a03706f7312106368616e6765436f6d6d697373696f6e180a20c0843d2a2065984ff3e6fe0d161a0a118c4e5d181a23c18f1d4bf59c78d178b7fcf0cadead320308c03e3a40f4e53ed1aaf56f878ab7cc13514164bbbe403dd52ac558ab88afdf8bc4829fd959aaf913ccf52e755185c681e13f84b4aaa723709faad7bc5de818fa96e48804
----
--

> =====

Reviewer 2: This step could be skipped, by adding a `--send` flag to the `transaction:create` command

+
. Upon successful execution of the transaction, the validator's commission value will be updated.
To check the updated values, invoke the xref:{url_getvalidator}[pos_getValidator] endpoint.
+
[source,json]
----
{
    "name": "web3lord",
    "totalStake": "2011000000000",
    "selfStake": "1011000000000",
    "lastGeneratedHeight": 4260,
    "isBanned": false,
    "reportMisbehaviorHeights": [],
    "consecutiveMissedBlocks": 0,
    "commission": 8500,
    "lastCommissionIncreaseHeight": 271,
    "sharingCoefficients": [ //This object will automatically have a value, once the validator starts sharing rewards.
        {
            "tokenID": "1234567800000000",
            "coefficient": "0b67f3fb9ef253f78c1b2c"
        }
    ],
    "address": "lskm87us5hykopm2f2nxa92z5ftbr9r52kg5b45e6",
    "punishmentPeriods": []
}
----

== Claiming rewards
After each block generated by the validator, both the validator and the staker(s) will earn a set percentage of rewards.

> For validators, the Lisk Protocol automatically increments the `lockedBalance` and `availableBalance` of the validator's account.

Reviewer 2: Why does it increment `lockedBalance` of the validator?

Reviewer 2: Does this only count for the commission, or also from the rewards that the validator receives from the self-stake?

Reviewer 2: Once @gkoumout has confirmed it, it should be added to the guide, as a note or something, imho it is useful information for stakers, that in certain situation, they will claim rewards automatically.

Concerning the locked balance, it would be nice if it is explained more in the guide how it works.

```suggestion
For validators, the Lisk Protocol automatically increments the `lockedBalance` and `availableBalance` of the validator's account.
The `availableBalance` in incremented by rewards corresponding to the commission and self-stake, the rest of the rewards belongs to validators and stakers according to their staked amount and is added to the `lockedBalance` of the validators, until the stakers claim the rewards.
```

Reviewer: The information that Sameer gave above is correct. This logic is described [here](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0071.md#after-transactions-execution).

Reviewer: I will add a suggestion from my side in a few minutes

Reviewer: @sameersubudhi About the case of automatic distribution it is not exactly as you described. In particular:

 - In case you (re) stake with a validator (either increase or decrease staked amount), indeed the rewards are assigned automatically to you. 
 - However, in case commission changes, nothing happens from the side of stakers.

Reviewer: I am not sure where is the best place to write this. Perhaps at the end, since in the very beginning it might be distracting from the context.

----
lisk-core transaction:create pos claimRewards 1000000 --json --pretty
----
--
Sidechain Node::
+
--
[source,bash]
----
./bin/run transaction:create pos claimRewards 1000000 --json --pretty
----
--
=====
+
. The *staker* should enter their `passphrase` to authenticate. 
+
---- 
? Please enter passphrase:  [hidden]
---- 
+
. The `claimRewards` command will return the transaction in the HEX string and JSON format.
+
[source,json]
---- 
{
  "transaction": "0a03706f73120c636c61696d52657761726473180720c0843d2a20145eb22910d905ba51a9ee2d0066ae97bebbff86084ce57f3df446b0cdd7942632003a40d5148b0c70efea8edeb6cd99ed0df23d72c1ec4845f796e8de510f30e56bd0ec7c20f744a4c4211040382396715a667a2d1e8ed3655020d844d1ba333db1ed01"
}
{
  "transaction": {
    "module": "pos",
    "command": "claimRewards",
    "fee": "1000000",
    "nonce": "7",
    "senderPublicKey": "145eb22910d905ba51a9ee2d0066ae97bebbff86084ce57f3df446b0cdd79426",
    "signatures": [
      "d5148b0c70efea8edeb6cd99ed0df23d72c1ec4845f796e8de510f30e56bd0ec7c20f744a4c4211040382396715a667a2d1e8ed3655020d844d1ba333db1ed01"
    ],
    "params": {},
    "id": "3da8f7278713d33fabe66a4da50cdbdcee674abc35ea0c12be4c65e0c162adf6"
  }
}
----
+

> . xref:{url_posting_transaction}[Post the transaction] to the node for execution and once the block is executed, check your xref:{url_getbalance}[token_getBalance] endpoint.

Reviewer 2: As above, this step could be skipped, by adding a `--send` flag to the `transaction:create` command

```suggestion
. Once the transaction is executed, check your xref:{url_getbalance}[token_getBalance] endpoint.
```
